Popup live toggle

siikamiika opened the pull request:
Fixes https://github.com/FooSoft/yomichan/issues/431

siikamiika commented:
There should probably be a forced settings update when using the root frame popup after using it from an iframe, but I haven't investigated it yet

toasted-nutbread commented on ext/fg/js/popup-nested.js:
```diff
@@ -53,3 +36,40 @@ async function popupNestedInitialize(id, depth, parentFrameId, url) {
+let popupNestedInitialized = false;
+
+async function popupNestedInitialize(id, depth, parentFrameId, url) {
+    if (popupNestedInitialized) {
```
I don't think `popupNestedInitialized` is necessary anymore. Please remove.

siikamiika replied:
It's called from `float.js`, but I can't think of cases when it would actually be called multiple times except maybe reloading the extension, and this check wouldn't help with that. Maybe there should be a check similar to the one in `search-frontend.js`

```javascript
        const node = document.querySelector(`script[src='${src}']`);
        if (node !== null) { continue; }
```

toasted-nutbread replied:
Yeah, that prepare should only be called once due to the `_prepareInvoked` guard, so that should handle it.

siikamiika replied:
https://github.com/FooSoft/yomichan/pull/439/commits/98d7e095a5c3533a9a2d787f3ee15b1640f77da0

toasted-nutbread commented on ext/fg/js/frontend-initialize.js:
```diff
@@ -25,49 +25,99 @@
+
+    const isIframe = !proxy && (window !== window.parent);
+
+    const initEventDispatcher = new EventDispatcher();
```
The use of `initEventDispatcher ` can probably be removed by some simplification.

`initEventDispatcher` is only used by `frontend`, which is only instantiated once. Add a `setPopup` method on it for `popupChange`, and `setDisabledOverride` can probably be handled in a similar way, or by setting the popup to `null`.

siikamiika replied:
https://github.com/FooSoft/yomichan/pull/439/commits/44ff4d889942987adf365a701c01eb8c3f98c2dd

I'd rather not set popup to `null` because then I'd have to potentially handle nulls all the places where popup is used.

toasted-nutbread commented on ext/mixed/js/text-scanner.js:
```diff
@@ -23,11 +23,13 @@
 
 class TextScanner {
-    constructor(node, ignoreElements, ignorePoints) {
+    constructor(node, ignoreElements, ignorePoints, canEnable=null) {
```
Rather than adding `canEnable`, might be better to add it as an argument to `setEnabled`. `setEnabled` is only called from `TextScanner.setOptions`, which is called from the subclasses of `TextScanner` (i.e. `Frontend`).

Meaning you can just pass `this.popup.depth <= this.options.scanning.popupNestingMaxDepth && !this._disabledOverride` here:
https://github.com/FooSoft/yomichan/blob/791c51f836c58ea7783acfc077d315c9897f224c/ext/fg/js/frontend.js#L147

And `true` here (or default value):
https://github.com/FooSoft/yomichan/blob/791c51f836c58ea7783acfc077d315c9897f224c/ext/bg/js/search-query-parser.js#L109

This should give better control over when the object is enabled/disabled.

siikamiika replied:
https://github.com/FooSoft/yomichan/pull/439/commits/6b47f80f195954f158de44b2971f653c3e23b7b8

siikamiika commented:
Seems like something's up with the CI, none of the tests go through

toasted-nutbread commented:
> Seems like something's up with the CI, none of the tests go through

Re-running it manually seems to have worked. ¯\\\_(ツ)\_/¯

siikamiika commented:
> There should probably be a forced settings update when using the root frame popup after using it from an iframe, but I haven't investigated it yet

Actually, the settings should be the same regardless of the iframe proxy because the context is the same there with the current profile conditions.

siikamiika commented on ext/fg/js/frontend-initialize.js:
```diff
@@ -24,49 +24,100 @@
-                if (action === 'rootPopupInformation') {
-                    resolve(params);
-                }
+    const {id, depth=0, parentFrameId, url=window.top.location.href, proxy=false, isSearchPage=false} = data;
```
I wonder what's going to happen with the line below:
```javascript
    const isIframe = !proxy && (window !== window.parent);
```

siikamiika replied:
I tested and that doesn't throw an error

siikamiika replied:
It seems that there's still an issue with `DisplayFloat`'s `optionsContext`. Fixing that would be quite complicated, and not necessarily something I want to do in this PR. Should I revert https://github.com/FooSoft/yomichan/pull/439/commits/b80fd00511c8d4fb02d0462c7f41e2c12c036367 and https://github.com/FooSoft/yomichan/pull/439/commits/b0d94e8b50c7a434415610175f63bd7d59e991d1 and create a new commit  making the default value `window.location.href`?

toasted-nutbread replied:
Makes sense to me.

toasted-nutbread commented on ext/fg/js/frontend-initialize.js:
```diff
@@ -24,49 +24,100 @@
+}
+
+async function getOrCreatePopup(depth) {
+    const frameOffsetForwarder = new FrameOffsetForwarder();
```
Two FrameOffsetForwarder instances can exist at the same time this way; one from `createIframePopupProxy` and one from `getOrCreatePopup`.

siikamiika replied:
https://github.com/FooSoft/yomichan/pull/439/commits/362ac84e6c6a0a4c2353232a5c44e1b2157de44d

toasted-nutbread commented:
Noticed an issue:

0. Open Chrome.
1. Open [test-document2.html](https://github.com/FooSoft/yomichan/blob/master/test/data/html/test-document2.html) in a new tab.
2. Open a popup by scanning content in test 1.
3. Open the options tab.
4. Change the value of **Maximum number of additional popups** or **Show iframe popups in root frame**.
5. test-document2.html hijacks the tab focus from the options tab.

toasted-nutbread commented:
Another issue I noticed, which may be a timing issue. Not as critical, but not sure how complex this would be to fix.

1. Change **Maximum number of additional popups** to some large value like 10.
2. Open a document, scan text, open all popups.
3. Change value of **Maximum number of additional popups** to 1. All popups will remain open.
4. Change value of **Maximum number of additional popups**  to 2. Most of the popups will now close.

The options will use the old value of `popupNestingMaxDepth`.

For testing, on line 290 I added:
```js
console.log('_canEnable', this.options, this.popup.depth, '<=', this.options !== null ? this.options.scanning.popupNestingMaxDepth : null, this._disabledOverride);
```

Logged this:
```
// Initially opening all popups
frontend.js:290 _canEnable null 0 <= null false
frontend.js:290 _canEnable null 1 <= null false
frontend.js:290 _canEnable null 2 <= null false
frontend.js:290 _canEnable null 3 <= null false
frontend.js:290 _canEnable null 4 <= null false
frontend.js:290 _canEnable null 5 <= null false
frontend.js:290 _canEnable null 6 <= null false
frontend.js:290 _canEnable null 7 <= null false
frontend.js:290 _canEnable null 8 <= null false
frontend.js:290 _canEnable null 9 <= null false
frontend.js:290 _canEnable null 10 <= null false
// Changed value of popupNestingMaxDepth to 1
frontend.js:290 _canEnable {…} 0 <= 10 false
frontend.js:290 _canEnable {…} 1 <= 10 false
frontend.js:290 _canEnable {…} 7 <= 10 false
frontend.js:290 _canEnable {…} 6 <= 10 false
frontend.js:290 _canEnable {…} 4 <= 10 false
frontend.js:290 _canEnable {…} 2 <= 10 false
frontend.js:290 _canEnable {…} 5 <= 10 false
frontend.js:290 _canEnable {…} 3 <= 10 false
frontend.js:290 _canEnable {…} 10 <= 10 false
frontend.js:290 _canEnable {…} 8 <= 10 false
frontend.js:290 _canEnable {…} 9 <= 10 false
// Changed value of popupNestingMaxDepth to 2
frontend.js:290 _canEnable {…} 0 <= 1 false
frontend.js:290 _canEnable {…} 1 <= 1 false
frontend.js:290 _canEnable {…} 7 <= 1 false
frontend.js:290 _canEnable {…} 6 <= 1 false
frontend.js:290 _canEnable {…} 4 <= 1 false
frontend.js:290 _canEnable {…} 2 <= 1 false
frontend.js:290 _canEnable {…} 5 <= 1 false
frontend.js:290 _canEnable {…} 3 <= 1 false
frontend.js:290 _canEnable {…} 10 <= 1 false
frontend.js:290 _canEnable {…} 8 <= 1 false
frontend.js:290 _canEnable {…} 9 <= 1 false
```

siikamiika commented:
> Noticed an issue:
> 
>     1. Open Chrome.
> 
>     2. Open [test-document2.html](https://github.com/FooSoft/yomichan/blob/master/test/data/html/test-document2.html) in a new tab.
> 
>     3. Open a popup by scanning content in test 1.
> 
>     4. Open the options tab.
> 
>     5. Change the value of **Maximum number of additional popups** or **Show iframe popups in root frame**.
> 
>     6. test-document2.html hijacks the tab focus from the options tab.
> 
> 
> Appears to have something to do with the iframes, since it doesn't occur on documents without iframes.

This happens when https://github.com/FooSoft/yomichan/pull/439/files#diff-c0a35793e5728d9d3e4b1301d23a4c3bR113-R115 is run because `Frontend.setPopup` calls `       this.onSearchClear(true);`. I can change it to `false`, but it could lead to hiding a popup so that it's still focused. I guess forcefully focusing the tab is worse. I remember we had a similar issue in the past, but it was fixed somehow.

siikamiika commented:
> Another issue I noticed, which may be a timing issue. Not as critical, but not sure how complex this would be to fix.
> 
>     1. Change **Maximum number of additional popups** to some large value like 10.
> 
>     2. Open a document, scan text, open all popups.
> 
>     3. Change value of **Maximum number of additional popups** to 1. All popups will remain open.
> 
>     4. Change value of **Maximum number of additional popups**  to 2. Most of the popups will now close.
> 
> 
> The options will use the old value of `popupNestingMaxDepth`.
> 
> For testing, on line 290 I added:
> 
> ```js
> console.log('_canEnable', this.options, this.popup.depth, '<=', this.options !== null ? this.options.scanning.popupNestingMaxDepth : null, this._disabledOverride);
> ```
> 
> Logged this:
> 
> ```
> // Initially opening all popups
> frontend.js:290 _canEnable null 0 <= null false
> frontend.js:290 _canEnable null 1 <= null false
> frontend.js:290 _canEnable null 2 <= null false
> frontend.js:290 _canEnable null 3 <= null false
> frontend.js:290 _canEnable null 4 <= null false
> frontend.js:290 _canEnable null 5 <= null false
> frontend.js:290 _canEnable null 6 <= null false
> frontend.js:290 _canEnable null 7 <= null false
> frontend.js:290 _canEnable null 8 <= null false
> frontend.js:290 _canEnable null 9 <= null false
> frontend.js:290 _canEnable null 10 <= null false
> // Changed value of popupNestingMaxDepth to 1
> frontend.js:290 _canEnable {…} 0 <= 10 false
> frontend.js:290 _canEnable {…} 1 <= 10 false
> frontend.js:290 _canEnable {…} 7 <= 10 false
> frontend.js:290 _canEnable {…} 6 <= 10 false
> frontend.js:290 _canEnable {…} 4 <= 10 false
> frontend.js:290 _canEnable {…} 2 <= 10 false
> frontend.js:290 _canEnable {…} 5 <= 10 false
> frontend.js:290 _canEnable {…} 3 <= 10 false
> frontend.js:290 _canEnable {…} 10 <= 10 false
> frontend.js:290 _canEnable {…} 8 <= 10 false
> frontend.js:290 _canEnable {…} 9 <= 10 false
> // Changed value of popupNestingMaxDepth to 2
> frontend.js:290 _canEnable {…} 0 <= 1 false
> frontend.js:290 _canEnable {…} 1 <= 1 false
> frontend.js:290 _canEnable {…} 7 <= 1 false
> frontend.js:290 _canEnable {…} 6 <= 1 false
> frontend.js:290 _canEnable {…} 4 <= 1 false
> frontend.js:290 _canEnable {…} 2 <= 1 false
> frontend.js:290 _canEnable {…} 5 <= 1 false
> frontend.js:290 _canEnable {…} 3 <= 1 false
> frontend.js:290 _canEnable {…} 10 <= 1 false
> frontend.js:290 _canEnable {…} 8 <= 1 false
> frontend.js:290 _canEnable {…} 9 <= 1 false
> ```

Caused by calling `_canEnable` before it got the newest options. Fixed in https://github.com/FooSoft/yomichan/pull/439/commits/b6f7f8c02637cef9cadd4ab2b9f253c42903aee6
